### PR TITLE
Eliminate need for trinity specific endpoint subclasses

### PR DESCRIPTION
**Pull request by pipermerriam**
### What was wrong?

The trinity specific subclasses of the `lahja.Endpoint` needed to be re-thought to make room for the upcoming `trio` base endpoints.  

### How was it fixed?

Rather than end up with versions for both `trio` and `asyncio` it seemed like the right type of thing for our `Service` abstraction to handle.  This resulted in a nice cascading cleanup of the boot logic as well.

Plugins take a little longer to boot now since they end up waiting for the endpoint to be ready.  I expect to clean this up further in the near term future so this shouldn't necessarily be viewed as the end state but rather a minimal step towards further plugin refactoring and cleanup.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Clean up commit history

[//]: # (For important changes, please add a new entry to the running release notes PR)
[//]: # (You can find the current one using: https://github.com/ethereum/trinity/pulls?q=is%3Aopen+is%3Apr+label%3A%22Release+Notes%22 )
[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [ ] Add entry to the [release notes PR](https://github.com/ethereum/trinity/pulls?q=is%3Aopen+is%3Apr+label%3A%22Release+Notes%22)

#### Cute Animal Picture

Cute Baby Hippo (5)

### CODE REVIEW

**Comment by cburgdorf** [comment]
This looks like a really nice cleanup coming up :+1: :tada:

File: trinity/event_bus.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+    async def _run(self) -> None:
+        async with AsyncioEndpoint.serve(self._connection_config) as endpoint:
+            self._endpoint = endpoint
+            # signal that the endpoint is now available
+            self._endpoint_available.set()
+
+            # run background task that automatically connects to newly announced endpoints
+            self.run_daemon_task(self._auto_connect_new_announced_endpoints())
+
+            # connect to the *main* endpoint which communicates enformation
```
Comment (cburgdorf) [inline trinity/event_bus.py]: typo **i**nformation

File: trinity/extensibility/asyncio.py [hunk 2]
```diff
@@ -1,53 +1,45 @@
...
 from trinity.extensibility.events import PluginStartedEvent
-from trinity.endpoint import TrinityEventBusEndpoint
 
 from .plugin import BaseIsolatedPlugin
 
 
 class AsyncioIsolatedPlugin(BaseIsolatedPlugin):
-    _event_bus: TrinityEventBusEndpoint = None
+    _event_bus: AsyncioEndpoint = None
+    _loop: asyncio.AbstractEventLoop
```
Comment (cburgdorf) [inline trinity/extensibility/asyncio.py]: Nitpick: I was confused why this turned into an instance member when it *seemed* to only be used within `_spawn_start` but now that I looked into it it looks like the `BeamChainExecutionPlugin` calls `self._loop` so maybe just drop the `_` and expose it as `loop` to get rid of that confusion.

File: trinity/plugins/builtin/ethstats/plugin.py [hunk 3]
```diff
@@ -125,5 +125,8 @@ def do_start(self) -> None:
             self.stats_interval,
         )
 
-        asyncio.ensure_future(exit_with_endpoint_and_services(self.event_bus, service))
+        asyncio.ensure_future(exit_with_services(
+            service,
+            self._event_bus_service,
```
Comment (cburgdorf) [inline trinity/plugins/builtin/ethstats/plugin.py]: Ah, ok. Now I see why we don't need the other `exit_with_*` API anymore

File: trinity/bootstrap.py [hunk 4]
```diff
@@ -96,25 +89,18 @@
...
 
 
 def main_entry(trinity_boot: BootFn,
                app_identifier: str,
-               plugins: Iterable[Type[BasePlugin]],
-               sub_configs: Iterable[Type[BaseAppConfig]]) -> None:
-
-    main_endpoint = TrinityMainEventBusEndpoint(name=MAIN_EVENTBUS_ENDPOINT)
+               plugins: Tuple[Type[BasePlugin], ...],
+               sub_configs: Sequence[Type[BaseAppConfig]]) -> None:
```
[the remark this answers is not in the document]
Comment (cburgdorf) [reply trinity/bootstrap.py]: I just learned that there's an actual [Robustness principle](https://en.wikipedia.org/wiki/Robustness_principle) that kinda maps nicely to this.

>Be conservative in what you send, be liberal in what you accept

[the remark this answers is not in the document]
Comment (pipermerriam) [reply trinity/bootstrap.py]: Accepting `Iterable` means accepting generators which feel like foot-guns, hence preferring `Sequence`.  I just hit a generator bug a few days ago that was frustratingly hard to track down.

Thoughts?  I'm :+1: on the idea, but we have rarely chosen to use generators in function APIs  due to the ease of mis-handling them.

[the remark this answers is not in the document]
Comment (cburgdorf) [reply trinity/bootstrap.py]: Mmmh..that's a valid point. Yep, `Sequence` seems fine then :+1:

File: tests/integration/test_trinity_cli.py [hunk 5]
```diff
@@ -243,9 +246,13 @@ def contains_substring(iterable, substring):
     stderr_logs = []
 
     # Collect logs up to the point when the sync begins so that we have enough logs for assertions
+    marker_seen_at = 0
     async for line in async_process_runner.stderr:
-        if "FastChainBodySyncer" in line:
+        if marker_seen_at != 0 and time.time() - marker_seen_at > 3:
```
[the remark this answers is not in the document]
Comment (pipermerriam) [reply tests/integration/test_trinity_cli.py]: Things are now starting up a little differently with things having to wait for their endpoint to be ready so things were taking longer to start and the `FastChainBodySyncer` marker was no longer sufficiently late in the process.  This was the best I could come up with.

**Pull request merged**